File: searching-algorithms/jump_search.js
"use strict";

const jumpSearch = (searchNumber, inputArray) => {
  try {
    console.log("jumpSearch");

    for (let i = 0; i < inputArray.length; i++) {
      if (i == 0) {
        continue;
      } else {
        if (inputArray[i - 1] > inputArray[i]) {
          throw "Array is not sorted in ascending order";
        }
      }
    }

    let jumpSize = Math.floor(Math.sqrt(inputArray.length));
    let initIndex = 0;
    let searchIndex = jumpSize - 1;
    let numberIndex = -1;

    while (true) {
      if (searchNumber > inputArray[searchIndex]) {
        initIndex = searchIndex;
        searchIndex = searchIndex + jumpSize;
        searchIndex = Math.min(searchIndex, inputArray.length - 1);

        if (searchIndex == initIndex) {
          break;
        }

        console.log("initIndex : ", initIndex);
        console.log("searchIndex : ", searchIndex);
      } else {
        for (let j = initIndex; j < searchIndex + 1; j++) {
          if (inputArray[j] == searchNumber) {
            numberIndex = j;
            break;
          }
        }

        break;
      }
    }

    return numberIndex;
  } catch (err) {
    console.log("jumpSearch error : ", err);
  }
};

let inputArray = [11, 3, 4, 11, 13];
let searchNumber = 13;

let numberIndex = jumpSearch(searchNumber, inputArray);
console.log("numberIndex : ", numberIndex);

File: searching-algorithms/jump_search.py
import math

def is_list_sorted(input_list):
    sorted_lst = True
    for idx, each_item in enumerate(input_list):
        if idx == 0:
            continue
        else:
            current_item = each_item
            previous_item = input_list[idx - 1]
            if current_item < previous_item:
                sorted_lst = False
                break

    return sorted_lst

def run_jump_search(search_number, input_list):

    if not is_list_sorted(input_list):
        raise Exception("List is not sorted in ascending order")
    
    number_index = -1
    jump_size = math.floor(math.sqrt(len(input_list))) 
    search_index = jump_size - 1
    init_index = 0

    while True:

        if search_number > input_list[search_index]:
            init_index = search_index
            search_index = search_index + jump_size
            search_index = min(search_index, len(input_list) - 1)

            if init_index == search_index:
                break

        else:
            # perform linear search 
            print("perform linear search")
            print(range(init_index, search_index + 1))

            for idx in range(init_index, search_index + 1):
                print(idx, input_list[idx])
                if input_list[idx] == search_number:
                    number_index = idx
                    break

            break

    return number_index


if __name__ == "__main__":
    search_number = 12
    input_list = [3, 4, 10, 11]
    number_index = run_jump_search(search_number, input_list)
    print("number found in index : ", number_index)
